### PR TITLE
feat: add makeRegression function

### PR DESCRIPTION
**Pull request by LuanSilveiraSouza**
From Issue #50

### CODE REVIEW

File: src/datasets/makeLowRankMatrix.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        .mul(tf.fill(singularIndexByRank.shape, -1))
+    )
+    .mul(tf.fill(singularIndexByRank.shape, 1 - tailStrength))
+  let tail = tf
+    .exp(singularIndexByRank.mul(tf.fill(singularIndexByRank.shape, -0.1)))
+    .mul(tf.fill(singularIndexByRank.shape, tailStrength))
+
+  const identity = tf.eye(n)
+  // FIX: Breaking here "Error: Operands could not be broadcast together with shapes x and y"
+  let s = identity.mul(lowRank.add(tail))
```
Comment (DirkToewe) [inline src/datasets/makeLowRankMatrix.ts:42]: You could also use `tf.diag( lowRank.add(tail) )`

File: src/datasets/makeLowRankMatrix.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  return tf.dot(tf.dot(u, s), tf.einsum('ij>ji', v)) as Tensor2D
```
Comment (DirkToewe) [inline src/datasets/makeLowRankMatrix.ts:44]: Unfortunately, `tf.einsum` is not yet supported in `tfjs-node` (As far as i know), but You could use:
```js
tf.matMul(tf.dot(u, s), v, /*transposeLeft=*/false, /*transposeRight=*/true)
```

File: src/datasets/makeLowRankMatrix.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  // Index of the singular values
+  let singularIndex = tf.linspace(0, n, n + 1)
+
+  // Build the singular profile by assembling signal and noise components
+  const singularIndexByRank = singularIndex.div(effectiveRank)
+  let lowRank = tf
+    .exp(
+      singularIndexByRank
+        .mul(singularIndexByRank)
+        .mul(tf.fill(singularIndexByRank.shape, -1))
```
Comment (DirkToewe) [inline src/datasets/makeLowRankMatrix.ts:33]: Alternatively, `.mul(-1)` or `.neg()` would also work (TFJS Numpy-style broadcasting)

File: src/datasets/makeRegression.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    [numberInformative, nFeatures - numberInformative],
+    0
+  )
+  let groundTruthRandom = tf.randomNormal([numberInformative, nTargets])
+  groundTruthRandom = groundTruthRandom.mul(
+    tf.fill(groundTruthRandom.shape, 100)
+  )
+  groundTruthA = groundTruthA.add(groundTruthRandom)
+
+  let groundTruth = tf.concat([groundTruthA, groundTruthB])
```
Comment (DirkToewe) [inline src/datasets/makeRegression.ts:60]: Alternative suggestion:
```js
let groundTruth = tf.concat([
  tf.randomNormal([numberInformative, nTargets], /*mean=*/0, /*stdDev=*/100),
  tf.zeros([nFeatures - numberInformative])
])
```

File: src/datasets/makeRegression.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  let Y: Tensor1D = tf.einsum('i,i->', X, groundTruth).as1D()
```
Comment (DirkToewe) [inline src/datasets/makeRegression.ts:62]: Tensor | Shape
----------|----------
X | [nSamples, nFeatures]
groundTruth | [nFeatures, nTargets]
Y | [nSamples, nTargets]


Should this not be `tf.einsum('ij,jk->', X, groundTruth)` or better `tf.matMul(X, groundTruth)` ?

File: src/datasets/makeRegression.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  effectiveRank?: number | null
+  tailStrength?: number
+  noise?: number
+  shuffle?: boolean
+  coef?: boolean
+}
+
+type MakeRegressionOutput =
+  | [Scikit2D, Scikit1D]
+  | [Scikit2D, Scikit1D, Scikit1D]
```
Comment (DirkToewe) [inline src/datasets/makeRegression.ts:21]: `Scikit1D` and `Scikit2D` are meant as input types which are as wide as possible. But as return types, `Tensor1D` and `Tensor2D` might be more practical (allows You to call methods like `X.print()` on them).

Also if `nTargets != 1`, then the return type would be `[Tensor2D, Tensor2D]` and `[Tensor2D, Tensor2D, Tensor2D]` respectively.

File: src/datasets/makeRegression.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    // Build the singular profile by assembling signal and noise components
+    const singularIndexByRank = singularIndex.div(effectiveRank)
+    const lowRank = tf
+      .exp(singularIndexByRank.square().neg())
+      .mul(1 - tailStrength)
+    let tail = tf.exp(singularIndexByRank.mul(-0.1)).mul(tailStrength)
+
+    let s = tf.diag(lowRank.add(tail))
+
+    return u.dot(s).dot(v.transpose()) as tf.Tensor2D
```
Comment (DirkToewe) [inline src/datasets/makeRegression.ts:108]: Thinking about it again, wouldn't the following also work and avoid one expensive `O(nSamples * n²)` matrix multiplication:
```js
let s = lowRank.add(tail)

return u.mul(s).dot(v.transpose()) as tf.Tensor2D
```

**Review by DirkToewe — APPROVED** [review]

File: src/datasets/makeRegression.test.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    expect(X.dot(model as Tensor).dataSync()).toEqual(y.dataSync())
+  })
+  it('test low rank matrix', () => {
+    let X = makeLowRankMatrix({
+      nSamples: 10,
+      nFeatures: 20
+    })
+    expect(X.shape).toEqual([10, 20])
+  })
+})
```
[the remark this answers is not in the document]
Comment (dcrescim) [reply src/datasets/makeRegression.test.ts:32]: Totally agree. I'll add a LinearRegression test.

**Pull request merged**